File: Week 3 & 4/04. Topics/ros2_ws/src/pkg_cpp_py/src/robot_news_station.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

using namespace std::chrono_literals;

class RobotNewsStation : public rclcpp::Node
{
public:
    RobotNewsStation() : Node("robot_news_station")
    {
        publisher_ = this->create_publisher<example_interfaces::msg::String>("robot_news", 10);
        timer_ = this->create_wall_timer(1s, std::bind(&RobotNewsStation::publish_news, this));
        RCLCPP_INFO(this->get_logger(), "Robot news has been started.");
    }

private:
    void publish_news()
    {
        auto msg = example_interfaces::msg::String();
        msg.data = std::string("Hi, this is ") + robot_name_ + std::string(" from the Robot news station");
        publisher_->publish(msg);
    }
    std::string robot_name_ {"Loybot"};
    rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<RobotNewsStation>());
    rclcpp::shutdown();
    return 0;
}

File: Week 3 & 4/04. Topics/ros2_ws/install/pkg_cpp_py/lib/pkg_cpp_py/robot_news_station.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from example_interfaces.msg import String

class RobotNewsStation(Node):
    def __init__(self):
        super().__init__("robot_news_station")
        self.robot_name_ = "Loybot"
        self.publisher_ = self.create_publisher(String, "robot_news", 10)
        self.timer_ = self.create_timer(0.5, self.publish_news)
        self.get_logger().info("Robot news has just started.")

    def publish_news(self):
        msg = String()
        msg.data = f"Hi, here is the latest news brought to you by {self.robot_name_}."
        self.publisher_.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    robot_news = RobotNewsStation()
    rclpy.spin(robot_news)
    rclpy.shutdown()

if __name__ == '__main__':
    main()